DIPs/1NNN-RA.md

```diff
@@ -0,0 +1,409 @@
+
+To make this work, a new root of all classes(in our case `ProtoObject`) and the `Equals` interface are needed, as well as an implementation for a mixin template that provides the implementation for opEquals. Then the `C` class must inherit from them and it must contain the mixin template as a field:
+```D
+class C : ProtoObject, Equals
+{
+    int a;
+    this(int) @safe {}
+    mixin ImplementEquals;
+}
+```
+In druntime/object.d:
+```D
+class ProtoObject {  }
+
+interface Equals
+{
+    const @nogc nothrow pure @safe scope
+    int equals(scope const ProtoObject rhs);
+}
+
+mixin template ImplementEquals(M...)
+{
+    // Equals mixin template implementation
+}
+...
+```
+
+The final points are of crucial importance:
+*  must be backwards compatible. The introduction of ProtoObject must not break code
+* The user must be allowed to chose what methods he desires to implement
+* Root objects must work in attributed code without issues. Since we, sadly, can't predict the future and know if and what attributes and qualifiers will be available in the language, this is yet another argument to have a ProtoObject with no methods.
+
+## Prior Work
+As some of you already know, both Java and C# have an `Object` class as the superclass of all classes. For both languages, `Object`
+defines a set of default methods, with Java baking into `Object` much more than C# does.
+
+#### Java
+
+Java's `java.lang.Object` defines the following methods:
+
+| Method name                                                 | Description |
+| ----------------------------------------------------------- | ------------------------------------------------------------------------------------------- |
+| boolean equals(Object o);                                   | Gives generic way to compare objects|
+| Class getClass();                                           | The Class class gives us more information about the object|
+| int hashCode();                                             | Returns a hash value that is used to search objects in a collection|
+| void notify();                                              | Used in synchronizing threads|
+| void notifyAll();                                           | Used in synchronizing threads|
+| String toString();                                          | Can be used to convert the object to String|
+| void wait();                                                | Used in synchronizing threads|
+| protected Object clone() throws CloneNotSupportedException; | Return a new object that are exactly the same as the current object|
+| protected void finalize() throws Throwable;                 | This method is called just before an object is garbage collected |
+
```

Comment:
Anything referring to code in the table above should be in code font. Rule of thumb: if a word is not used with its natural language meaning, it should be in a different font.